examples/horizontal_convection.jl

```diff
@@ -0,0 +1,386 @@
+# as "horizontal convection".
+#
+# This example demonstrates:
+#
+#   * How to use `ComputedField`s for output
+#   * How to post-process saved output
+#
+# ## Install dependencies
+#
+# First let's make sure we have all required packages installed.
+
+# ```julia
+# using Pkg
+# pkg"add Oceananigans, JLD2, Plots"
+# ```
+
+# ## Horizontal convection
+#
+# We'll consider here the two-dimensional verion of horizontal convection in which a flow
+# ``\boldsymbol{u} = (u, w)`` on the ``(x, z)``-plane is evolved under the effect of gravity
+# and with a prescribed, non-uniform buoyancy at the top-surface of the domain.
+#
+# ```math
+# \begin{gather}
+# \partial_t \boldsymbol{u} + \boldsymbol{u \cdot \nabla} \boldsymbol{u} + \boldsymbol{\nabla}p = b \hat{\boldsymbol{z}} + \nu \nabla^2 \boldsymbol{u} \, , \\
+# \partial_t b + \boldsymbol{u \cdot \nabla} b = \kappa \nabla^2 b \, , \\
+# \boldsymbol{\nabla \cdot u} = 0 \, .
+# \end{gather}
+# ```
+# 
+# The domain here is ``-L_x/2 \le x \le L_x/2`` and ``-H \le z \le 0``.
+#
+# ### Boundary conditions
+#
+# At the surface, the imposed buoyancy is ``b_s(x, z = 0, t) = - b_* \cos(2 \pi x / L_x)``
+# while zero-flux boundary conditions are imposed on all other boundaries. We use free-slip 
+# boundary conditions on ``u`` and ``w`` at all boundaries.
+#
+# ### Non-dimensional control parameters
+#
+# The problem is characterized by three non-dimensional parameters. The first is the domain's
+# aspect ratio, ``A = L_x / H`` and the other two are the Rayleigh (``Ra``) and Prandtl (``Pr``)
+# numbers:
+#
+# ```math
+# Ra = \frac{L_x^3 b_*}{\nu \kappa} \, , \quad \text{and}\, \quad Pr = \frac{\nu}{\kappa} \, .
+# ```
+#
+# The Prandtl number expresses the ratio of momentum over heat diffusion; the Rayleigh number
+# is a measure of the relative importance of gravity over viscosity in the momentum equation.
+#
+# For a domain with a given extent, the nondimensional values of ``Ra`` and ``Pr`` uniquely
+# determine the viscosity and diffusivity, i.e.,
+# 
+# ```math
+# \nu = \sqrt{\frac{Pr b_* L_x^3}{Ra}} \quad \text{and} \quad \kappa = \sqrt{\frac{b_* L_x^3}{Pr Ra}} \, .
+# ```
+#
+# Now let's code these things up! We start off by importing `Oceananigans` and `Printf`.
+
+using Oceananigans
+using Printf
+
+# ## The grid
+#
+# We use a two-dimensional grid with an aspect ratio ``A = L_x / H = 2``.
+
+H = 1.0          # vertical domain extent
+Lx = 2H          # horizontal domain extent
+Nx, Nz = 128, 64 # horizontal, vertical resolution
+
+grid = RegularRectilinearGrid(size = (Nx, Nz),
+                                 x = (-Lx/2, Lx/2),
+                                 z = (-H, 0),
+                              halo = (3, 3),
+                          topology = (Bounded, Flat, Bounded))
+
+# ## Boundary conditions
+#
+# We impose a surface buoyancy boundary condition,
+#
+# ```math
+# b(x, z = 0, t) = - b_* \cos (2 \pi x / L_x) \, .
+# ```
+
+const b★ = 1.0
+const k = 2π / Lx
```

Comment:
I don't think `k` is used in the boundary condition or anywhere in the model.

Reply:
I'd like to pass it as a parameter.... --> https://github.com/CliMA/Oceananigans.jl/pull/1847#issuecomment-881780508

Reply:
Can you reference as a global? Using `const` implies you're doing that but I don't think that is actually happening.

Reply:
Are you suggesting
```
global b★ = 1.0
global k = 2π / Lx
```
?

Reply:
Doing the above changed gave the *exact same thing*:

```
┌ Warning: failed to run `@example` block in src/generated/horizontal_convection.md:221-223
│ ```@example horizontal_convection
│ run!(simulation)
│ ```
│   c.value = TypeError: in <:, expected Type, got a value of type JLD2.OnDiskRepresentation{(0,), Tuple{NamedTuple{(:b★, :k), Tuple{Float64, Float64}}}, Tuple{NamedTuple{(:b★, :k), Tuple{Float64, Float64}}}}
```

Reply:
I don't mind hard coding... but I feel I'm loosing faith in parametric construction of boundary conditions. Also in other examples seems to work, why not in this one?